### PR TITLE
added support for eip-2718 && 2930

### PR DESCRIPTION
**Pull request by malonaz**
## What was wrong?
The library does not support EIP-2718 (Typed Transactions) and EIP-2930 (Optional Access Lists).

Issue #114

## How was it fixed?
EIP-2718 throws away the well-understood transaction format, leaving individual new transaction types free to define their encoding format. I initially attempted to plumb this through the existing `transactions` type, but eventually decided to introduce a `typed_transactions` library. The philosophy is simple: A lot of complexity in eth transactions (v vs chain etc) exist because of the traditional transaction format. The introduction of the transaction types in EIP-2718 means we no longer have to do hacky +V_OFFSET etc. I found it hard to code around these in the existing transaction library without introducing more complexity to an already complex piece of code. In addition, the traditional encoding format (RLP) leaks in and out of functions within `transactions`, making it tricky to introduce a new encoding format.

### Cute Animal Picture
image

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Add tests

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/eth-account/blob/master/newsfragments/README.md)

[//]: # (See: https://eth-account.readthedocs.io/en/latest/contributing.html#pull-requests)
- [x] Add entry to the [release notes](https://github.com/ethereum/eth-account/blob/master/newsfragments/README.md)

### CODE REVIEW

**Comment by marcgarreau** [comment]
First, a big thank you for the work that's gone into this. I think all of us would-be reviewers are knee-deep in other issues at the moment, but someone will circle back as soon as we're able. These are nontrivial updates, so pinging @carver and @pipermerriam directly in case they want to take a peek.

**Comment by carver** [comment]
I took just a very quick look first, focusing on the API level, and it looks roughly like what I'd expect. Thanks for putting this together!

I'll do a deeper dive in a bit. The lint looks like an easy fix, just deleting the "type: ignore" comment.

**Comment by malonaz** [comment]
I tried deleting the three instances of 'type: ignore' comments, but it did not fix the issue.

**Comment by palkeo** [comment]
This PR would be the ideal foundation for adding the EIP-1559 transaction type.
I think it will be important to have before the London hard fork.

**Comment by lightclient** [comment]
@carver seems like the lint is failing on `master`. I fixed it in #116.

**Comment by kclowes** [comment]
I wasn't able to get to this today, but planning on doing a deep-dive review on Wednesday. Thanks for the PR @malonaz!

File: eth_account/_utils/transactions.py [hunk 1]
```diff
@@ -34,12 +31,22 @@
...
 )
 
-VALID_EMPTY_ADDRESSES = {None, b'', ''}
+from .typed_transactions import (
+    TypedTransaction
+)
 
 
 def serializable_unsigned_transaction_from_dict(transaction_dict):
+    if 'type' in transaction_dict:
```
Comment (palkeo) [inline eth_account/_utils/transactions.py:46]: Here you probably want to avoid entering the condition if there is a transaction type set, but it's 0?

Comment (malonaz) [reply eth_account/_utils/transactions.py:46]: the TypedTransaction.from_dict() method handles this behaviour. do you feel it's worth adding another check here?

**Comment by kclowes** [comment]
I started going through this today - like others have said it looks good from a high level! I'll continue going though tomorrow. One nitpicky thing I noticed is that you use `assert` a lot for validation. We prefer to `raise` instead, since it's easy for users to turn off `AssertionError`s. Thanks!

**Comment by malonaz** [comment]
I've switched to raise. thanks @kclowes

**Review by kclowes — COMMENTED** [review]
Thanks for your patience @malonaz! I finished going through and doing some testing and this looks solid to me. I only have two comments:
- Do you mind adding either a note or example to the docstrings for `sign_transaction` and `recover_transaction` in `eth_account/account.py` to show usage for the 2930 access list type? 
- I added one note in the PR about making an abstract class that will enforce that certain methods are present on any child `Transaction` classes. Let me know what you think or if that needs clarification!

File: eth_account/_utils/typed_transactions.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+            transaction=transaction,
+        )
+
+    def hash(self):
+        """
+        Hashes this TypedTransaction to prepare it for signing. As per the EIP-2718 specifications,
+        the hashing format is dictated by the transaction type itself, and so we delegate the call.
+        Note that the return type will be bytes.
+        """
+        return self.transaction.hash()
```
[the remark this answers is not in the document]
Comment (malonaz) [reply eth_account/_utils/typed_transactions.py:112]: @kclowes I've added an example to the `sign_transaction` method, but I'm not sure what I can do for the `recover_transaction` method. The example would look pretty much like the existing one? lmk if I've misunderstood something.

[the remark this answers is not in the document]
Comment (malonaz) [reply eth_account/_utils/typed_transactions.py:140]: @kclowes I've added the abstract class. lmk if this is what you had in mind.

[the remark this answers is not in the document]
Comment (kclowes) [reply eth_account/_utils/typed_transactions.py:144]: Yep, that is what I was thinking! Thanks!

**Comment by kclowes** [comment]
Thanks @malonaz, this is looking good to me! @carver - do you want to take another look?

**Comment by malonaz** [comment]
hey @kclowes / @carver any updates on this? Thanks

**Comment by kclowes** [comment]
@malonaz thank for the reminder! I have a few other things at the top of my priority queue but should get to it by Wednesday at the latest!

**Review by kclowes — APPROVED** [review]
Thanks @malonaz! I'll rebase and merge shortly!

**Comment by malonaz** [comment]
Awesome. Thanks

**Pull request merged**